pennylane/_grad.py

```diff
@@ -101,7 +94,25 @@ def _get_grad_fn(self, args):
     def __call__(self, *args, **kwargs):
         """Evaluates the gradient function, and saves the function value
         calculated during the forward pass in :attr:`.forward`."""
-        grad_value, ans = self._get_grad_fn(args)(*args, **kwargs)
+        if self._argnum is not None:
+            new_args = []
+
+            for idx, arg in enumerate(args):
+                if idx == self._argnum or (
+                    isinstance(self._argnum, Sequence) and idx in self._argnum
+                ):
+                    import pennylane as qml
+
+                    arg = qml.numpy.asarray(arg, requires_grad=True)
+
+                new_args.append(arg)
+
+            grad_value, ans = self._grad_with_forward(self._fun, argnum=self._argnum)(
+                *new_args, **kwargs
+            )
+        else:
+            grad_value, ans = self._get_grad_fn(args)(*args, **kwargs)
```

Comment:
@dime10 this is the only main change I had to make for some of the logic. I noticed that doing something like

```python
qml.grad(fn, arngum=0)(0.5)
```

wasn't working, when it _should_ work --- `argnum` is an 'override' that basically says 'treat argument 0 as trainable even if it has `requires_grad=False`'.

Here, if `argnum` is provided, I iterate through the argument lists and explicitly set `requires_grad=True` for that argument. I do the same below for Jacobian.

Reply:
I'd like to understand this issue a little better. Does that mean our argnum logic still requires the arguments selected with argnum to carry `requires_grad=True` internally?
In which case, I'm trying to figure out where the requires_grad attribute plays a role again. But when I run your example on the code before your changes, it seems to works fine:
```py
@qml.qnode(qml.device('default.qubit', wires=1))
def fn(x):
    qml.RY(x, wires=0)
    return qml.expval(qml.PauliZ(0))

print(qml.grad(fn, argnum=0)(0.5))
```
I believe since the parameter is converted to an ArrayBox, it is recognized as differentiable across our code. Would you have a more complete example where you ran into the problem of it not differentiating even with the argnum argument?

Reply:
I think I came across this issue in `test_classical_gradients.py`. Namely, it was doing something like this:

```python
>>> x_vals = np.linspace(-10, 10, 16, endpoint=False)
>>> g = qml.grad(np.sin, 0)
```

This will fail, because setting `argnum=0` will not alter the fact that `x_vals` has `requires_grad=False` now by default; I get the following error:

https://github.com/PennyLaneAI/pennylane/blob/1fd370d0ee2b14c3ddef9af6e5eccbf8784350a6/pennylane/numpy/tensor.py#L305-L307

Reply:
Ah I see, it works for scalars and vanilla numpy arrays, but for PennyLane tensors it does indeed fail. Nice catch!